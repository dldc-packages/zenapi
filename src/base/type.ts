import type { TEntityAny } from '../entity';
import type { TQuery } from '../query';
import { defineAbstract, defineType } from '../types';

const stringType = defineType<null, undefined>('string');
const dateType = defineType<null, undefined>('date');
const numberType = defineType<null, undefined>('number');
const booleanType = defineType<null, undefined>('boolean');
const jsonType = defineType<null, undefined>('json');
const nilType = defineType<null, undefined>('nil');
const enumType = defineType<readonly string[], undefined>('enum');

export type TNullableDef = { nullable: TQuery | false };

const nullableType = defineType<TEntityAny, TNullableDef>('nullable');

export type TEntityRecord = Record<string, TEntityAny>;

const objectType = defineType<TEntityRecord, string>('object');

export type TListDef =
  | { type: 'all'; select: TQuery }
  | { type: 'first'; select: TQuery }
  | {
      type: 'paginate';
      page: number | { page: number; pageSize: number };
      select: TQuery;
    };

const listType = defineType<TEntityAny, TListDef>('list');

export interface IInputDef {
  input: any;
  select: TQuery;
}

const inputType = defineType<TEntityAny, IInputDef>('input');

export const types = {
  string: stringType,
  date: dateType,
  number: numberType,
  boolean: booleanType,
  json: jsonType,
  nil: nilType,
  enum: enumType,
  nullable: nullableType,
  object: objectType,
  list: listType,
  input: inputType,
};

export type TAbstractObjectDef = Record<string, TQuery>;

const objectAbstract = defineAbstract<TAbstractObjectDef>('object');

const errorBoundaryAbstract = defineAbstract<TQuery>('errorBoundary');

export const abstracts = {
  object: objectAbstract,
  errorBoundary: errorBoundaryAbstract,
};
