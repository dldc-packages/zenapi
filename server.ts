export * from "@dldc/stack";

export * from "./src/server/context.ts";
export * from "./src/server/engine.ts";
export * from "./src/server/graph.ts";
export * from "./src/server/handlers.ts";
export * from "./src/server/parse.ts";
export * from "./src/server/structure.types.ts";
export * from "./src/server/types.ts";
export * from "./src/utils/types.ts";
