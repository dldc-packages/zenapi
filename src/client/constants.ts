export const RESULT = Symbol("RESULT");
export const TO_JSON = Symbol("TO_JSON");
