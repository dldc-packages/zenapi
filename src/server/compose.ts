import { composeAdvanced } from "@dldc/compose";
import type { TMiddleware } from "../../server.ts";
import type { ApiContext } from "./context.ts";

export function compose(...middlewares: TMiddleware[]): TMiddleware {
  return composeAdvanced<
    ApiContext,
    Promise<ApiContext> | ApiContext,
    Promise<ApiContext>
  >(async (v) => await v, middlewares);
}

export function withContext(
  mid: TMiddleware,
  update: (ctx: ApiContext) => ApiContext,
): TMiddleware {
  return async (ctx, next) => await mid(update(ctx), next);
}
