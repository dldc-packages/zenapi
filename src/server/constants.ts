export const PATH = Symbol("PATH");
export const STRUCTURE = Symbol("STRUCTURE");
export const SCHEMA = Symbol("SCHEMA");
export const GRAPH_KEY = Symbol("GRAPH_KEY");
export const GET = Symbol("GET");
// Used in schema path to indicate the root of the schema
export const ROOT = Symbol("ROOT");
export const TYPES = Symbol("TYPES");
