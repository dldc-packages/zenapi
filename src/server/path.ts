import type { TAllStructure } from "./structure.ts";
import type { TGraphRefBase } from "./types.ts";

export type TNormalizedPath = symbol[];

export type TPathCache = WeakMap<
  symbol,
  { ref: TGraphRefBase; structure: TAllStructure }
>;

// export function normalizePath(
//   cache: TPathCache,
//   schema: TSchemaAny,
//   rootStructure: TRootStructure,
//   path: TGraphPath,
// ): TNormalizedPath {
//   const normalized: TNormalizedPath = [];
//   const rootRef = schema.ref as TGraphRefBase;
//   let current: TGraphRefBase = rootRef;
//   let structure: TAllStructure = rootStructure;
//   while (path.length > 0) {
//     const item = path.shift()!;
//     const isRoot = current[GRAPH_KEY] === rootRef[GRAPH_KEY];
//     if (item === ROOT) {
//       if (isRoot) {
//         // Root while already at root, should we just skip?
//         throw new Error(`Invalid path`);
//       }
//       normalized.push(current[GRAPH_KEY]);
//       current = rootRef;
//       structure = rootStructure;
//       continue;
//     }
//     if (structure.kind === "root") {
//       if (!isRoot) {
//         throw new Error(`Invalid path`);
//       }
//       const subStructure: TStructure = structure.types[item];
//       if (!subStructure) {
//         throw new Error(`Invalid path: ${item} not found at root`);
//       }
//       structure = subStructure;
//       current = current[GET](item);
//       continue;
//     }
//     if (isRoot) {
//       throw new Error(`Invalid path`);
//     }
//     if (structure.kind === "object") {
//       const foundProp: TStructureObjectProperty | undefined = structure
//         .properties.find((p) => p.name === item);
//       if (!foundProp) {
//         throw new Error(`Invalid path: ${item} not found in ???`);
//       }
//       structure = foundProp.structure;
//       current = current[GET](item);
//       continue;
//     }
//     if (structure.kind === "ref") {
//       path.unshift(ROOT, structure.ref, item);
//       structure = rootStructure;
//       continue;
//     }
//     throw new Error(`Not implemented ${structure.kind}`);
//   }

//   if (current[GRAPH_KEY] !== rootRef[GRAPH_KEY]) {
//     normalized.push(current[GRAPH_KEY]);
//   }

//   return normalized;
// }
