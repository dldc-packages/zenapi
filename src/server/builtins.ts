import * as v from "@valibot/valibot";
import { TYPES } from "./constants.ts";
import { graphInternal, type TGraphBaseAny } from "./graph.ts";
import type { TPrepareContext, TQueryUnknown } from "./prepare.ts";
import type { TBuiltinStructure, TRootStructure } from "./structure.types.ts";
import type {
  TBuiltinsFromConfig,
  TGraphBuiltins,
  TMiddleware,
} from "./types.ts";

export type TBuiltinGetSchema = (
  context: TPrepareContext,
  graph: TGraphBaseAny,
) => v.BaseSchema<any, any, any>;

export type TBuiltinPrepare = (
  structure: TBuiltinStructure,
  graph: TGraphBaseAny,
  query: TQueryUnknown,
) => TMiddleware | null;

export type TBuiltinMatch = (
  graph: TGraphBaseAny,
  stucture: TBuiltinStructure,
  value: unknown,
) => boolean;

export interface TBuiltinConfig<T> {
  [TYPES]: T;
  /**
   * Schema used to validate input
   */
  getSchema: TBuiltinGetSchema;
  /**
   * Create middleware to resolve the value
   */
  prepare: TBuiltinPrepare;
  /**
   * Used to disambiguate the value when used in a union
   */
  match: TBuiltinMatch;
}

export type TBuiltinTypesConfig = Record<
  string,
  TBuiltinConfig<any> | null
>;

export type TBuiltinTypes = Record<string, TBuiltinStructure>;

export function createBuiltins<Conf extends TBuiltinTypesConfig>(
  config: Conf,
): TGraphBuiltins<TBuiltinsFromConfig<Conf>> {
  const configResolved = {
    ...DEFAULT_BUILTINS,
    ...config,
  };
  const builtins: TBuiltinStructure[] = [];
  for (const [name, config] of Object.entries(configResolved)) {
    if (config) {
      const { getSchema, prepare, match } = config;
      builtins.push({
        kind: "builtin",
        key: `builtin.${name}`,
        name,
        getSchema,
        prepare,
        match,
      });
    }
  }

  const rootStructure: TRootStructure = {
    kind: "root",
    key: "root",
    mode: "builtins",
    builtins,
    types: [],
  };

  return graphInternal({
    rootStructure,
    localTypes: {},
    path: [],
  }) as any;
}

export function builtin<T>(
  config: Omit<TBuiltinConfig<T>, typeof TYPES>,
): TBuiltinConfig<T> {
  return {
    [TYPES]: null as any,
    ...config,
  };
}

export type TDefaultBuiltins = {
  Date: TBuiltinConfig<Date>;
};

export const DEFAULT_BUILTINS: TDefaultBuiltins = {
  Date: builtin<Date>({
    getSchema: () => v.date(),
    prepare: (structure) => {
      return async (ctx, next) => {
        const res = await next(ctx);
        const value = res.value;
        if (value === undefined) {
          throw new Error(`Value is undefined at ${structure.key}`);
        }
        if (!(value instanceof Date)) {
          throw new Error(`Value is not a Date at ${structure.key}`);
        }
        return res;
      };
    },
    match: (_graph, _structure, value) => {
      if (value === undefined) {
        return false;
      }
      return value instanceof Date;
    },
  }),
};

export const DEFAULT_BUILTINS_GRAPH = createBuiltins<TDefaultBuiltins>(
  DEFAULT_BUILTINS,
);
