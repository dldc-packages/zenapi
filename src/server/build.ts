import * as v from "@valibot/valibot";
import { compose } from "./compose.ts";
import { GET, STRUCTURE } from "./constants.ts";
import { ApiContext } from "./context.ts";
import type { TGraphBase } from "./graph.ts";
import { getStructureMiddleware, getStructureValue } from "./structure.ts";
import type { TRootStructure } from "./structure.types.ts";
import type { TMiddleware } from "./types.ts";

export type TBuildFromOperator = (
  context: TBuildResolverContext,
  graph: TGraphBase,
  query: unknown,
) => TMiddleware;

interface TBuildResolverContext {
  rootGraph: TGraphBase;
  rootStructure: TRootStructure;
  operators: Record<string, TBuildFromOperator>;
}

const QueryObjectSchema = v.looseObject({ kind: v.string() });

export function build(
  context: TBuildResolverContext,
  graph: TGraphBase,
  query: unknown,
): TMiddleware {
  return buildFromOperator(context, graph, query);
}

export function buildFromOperator(
  context: TBuildResolverContext,
  graph: TGraphBase,
  query: unknown,
): TMiddleware {
  const queryObject = v.parse(QueryObjectSchema, query);
  const kind = queryObject.kind;
  const builder = context.operators[kind];
  if (!builder) {
    throw new Error(`Unknown operator kind: ${kind}`);
  }
  return builder(context, graph, query);
}

export function buildFromPath(
  context: TBuildResolverContext,
  graph: TGraphBase,
  path: (string | number)[],
): TMiddleware {
  const [prop, ...rest] = path;
  console.log("buildFromPath", prop, rest);
  if (!prop) {
    throw new Error("Path cannot be empty");
  }
  const subGraph = graph[GET](prop);
  const structure = subGraph[STRUCTURE];
  console.log(structure.kind, structure.key);
  const baseMid = getStructureMiddleware(context.rootStructure, structure);
  // TODO, use user middleware
  const mid = compose(baseMid);
  if (rest.length === 0) {
    return mid;
  }
  const subMid = buildFromPath(context, subGraph, rest);
  return async (ctx, next) => {
    const parentRes = await mid(ctx, (ctx) => Promise.resolve(ctx));
    console.log("parentRes", parentRes.value);
    const subValue = getStructureValue(structure, parentRes.value, prop);
    return subMid(
      parentRes.withValue(subValue).with(
        ApiContext.GraphKey.Provider(subGraph),
      ),
      next,
    );
  };
}
