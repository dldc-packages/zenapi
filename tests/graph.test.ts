import { assertEquals } from "@std/assert";
import { resolve } from "@std/path";
import { assertSnapshot } from "@std/testing/snapshot";
import { parseSchema, type TStructureObject } from "../server.ts";
import { PATH } from "../src/server/constants.ts";
import type { BasicTypes } from "./schemas/basic.types.ts";
import type { TodoListTypes } from "./schemas/todolist.types.ts";

Deno.test("todolist snapshot structure", async (test) => {
  const schema = parseSchema<TodoListTypes>(
    resolve("./tests/schemas/todolist.ts"),
  );
  await assertSnapshot(test, schema.structure);
});

const schema = parseSchema<BasicTypes>(resolve("./tests/schemas/basic.ts"));

const g = schema.graph;

Deno.test("snapshot structure", async (test) => {
  await assertSnapshot(test, schema.structure);
});

Deno.test("g.User.age", () => {
  const p1 = g.User.age;
  const path = p1[PATH];
  const userAgeStructure =
    (schema.structure.types.User as TStructureObject).properties.find((p) =>
      p.name === "age"
    )!.structure;
  assertEquals(path, [userAgeStructure]);
});

Deno.test("g.User.group", () => {
  const p2 = g.User.group;
  const path = p2[PATH];
  const userGroupStructure =
    (schema.structure.types.User as TStructureObject).properties.find((p) =>
      p.name === "group"
    )!.structure;
  assertEquals(path, [userGroupStructure]);
});

Deno.test("g.User.group.name", () => {
  const p3 = g.User.group.name;
  const path = p3[PATH];
  const userGroupStructure =
    (schema.structure.types.User as TStructureObject).properties.find((p) =>
      p.name === "group"
    )!.structure;
  const groupNameStructure =
    (schema.structure.types.Group as TStructureObject).properties.find((p) =>
      p.name === "name"
    )!.structure;
  assertEquals(path, [userGroupStructure, groupNameStructure]);
});

Deno.test("g.User.group.users.items.group", () => {
  const p4 = g.Graph.user.group.users.items.group;
  const path = p4[PATH];
  assertEquals(path.map((p) => p.key), [
    "Graph.user",
    "User.group",
    "Group.users.items",
    "User.group",
  ]);
});

Deno.test("g.User.maybeGroup", () => {
  const p = g.User.maybeGroup;
  const path = p[PATH];
  assertEquals(path.map((p) => p.key), [
    "User.maybeGroup",
  ]);
});

Deno.test("g.User.maybeGroup._(g.Group)", () => {
  const p = g.User.maybeGroup._(g.Group);
  const path = p[PATH];
  assertEquals(path.map((p) => p.key), [
    "User.maybeGroup",
    "Group",
  ]);
});

Deno.test("g.User.maybeGroup._(g.Group).name", () => {
  const p = g.User.maybeGroup._(g.Group).name;
  const path = p[PATH];
  assertEquals(path.map((p) => p.key), [
    "User.maybeGroup",
    "Group.name",
  ]);
});

// Deno.test("Simple graph", () => {
//   const p4 = g.User.group.users.items.group;
//   assertEquals(p4[PATH], [
//     userGroupStructure,
//     groupUserStructure,
//     userGroupStructure,
//   ]);
// });
