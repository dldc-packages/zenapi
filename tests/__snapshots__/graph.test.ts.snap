export const snapshot = {};

snapshot[`todolist snapshot structure 1`] = `
{
  key: "",
  kind: "root",
  types: {
    App: {
      key: "App",
      kind: "object",
      properties: [
        {
          name: "appName",
          optional: false,
          structure: {
            key: "App.appName",
            kind: "primitive",
            type: "string",
          },
        },
        {
          name: "todos",
          optional: false,
          structure: {
            items: {
              key: "App.todos.items",
              kind: "ref",
              params: undefined,
              ref: "Todo",
            },
            key: "App.todos",
            kind: "array",
          },
        },
      ],
    },
    CoinHeads: {
      key: "CoinHeads",
      kind: "object",
      properties: [
        {
          name: "foo",
          optional: false,
          structure: {
            key: "CoinHeads.foo",
            kind: "primitive",
            type: "string",
          },
        },
      ],
    },
    CoinTails: {
      key: "CoinTails",
      kind: "object",
      properties: [
        {
          name: "bar",
          optional: false,
          structure: {
            key: "CoinTails.bar",
            kind: "primitive",
            type: "string",
          },
        },
      ],
    },
    Config: {
      key: "Config",
      kind: "object",
      properties: [
        {
          name: "env",
          optional: false,
          structure: {
            key: "Config.env",
            kind: "object",
            properties: [
              {
                name: "version",
                optional: false,
                structure: {
                  key: "Config.env.version",
                  kind: "primitive",
                  type: "string",
                },
              },
              {
                name: "num",
                optional: false,
                structure: {
                  key: "Config.env.num",
                  kind: "primitive",
                  type: "number",
                },
              },
              {
                name: "str",
                optional: false,
                structure: {
                  key: "Config.env.str",
                  kind: "primitive",
                  type: "string",
                },
              },
              {
                name: "bool",
                optional: false,
                structure: {
                  key: "Config.env.bool",
                  kind: "primitive",
                  type: "boolean",
                },
              },
            ],
          },
        },
      ],
    },
    Graph: {
      key: "Graph",
      kind: "object",
      properties: [
        {
          name: "users",
          optional: false,
          structure: {
            key: "Graph.users",
            kind: "object",
            properties: [
              {
                name: "byId",
                optional: false,
                structure: {
                  arguments: {
                    arguments: [
                      {
                        name: "id",
                        optional: false,
                        structure: {
                          key: "Graph.users.byId.arguments.id",
                          kind: "primitive",
                          type: "string",
                        },
                      },
                    ],
                    key: "Graph.users.byId.arguments",
                    kind: "arguments",
                  },
                  key: "Graph.users.byId",
                  kind: "function",
                  returns: {
                    key: "Graph.users.byId.returns",
                    kind: "ref",
                    params: undefined,
                    ref: "User",
                  },
                },
              },
            ],
          },
        },
        {
          name: "apps",
          optional: false,
          structure: {
            key: "Graph.apps",
            kind: "object",
            properties: [
              {
                name: "all",
                optional: false,
                structure: {
                  arguments: {
                    arguments: [
                      {
                        name: "pagination",
                        optional: true,
                        structure: {
                          key: "Graph.apps.all.arguments.pagination",
                          kind: "ref",
                          params: undefined,
                          ref: "PaginationParams",
                        },
                      },
                    ],
                    key: "Graph.apps.all.arguments",
                    kind: "arguments",
                  },
                  key: "Graph.apps.all",
                  kind: "function",
                  returns: {
                    items: {
                      key: "Graph.apps.all.returns.items",
                      kind: "ref",
                      params: undefined,
                      ref: "App",
                    },
                    key: "Graph.apps.all.returns",
                    kind: "array",
                  },
                },
              },
              {
                name: "byId",
                optional: false,
                structure: {
                  arguments: {
                    arguments: [
                      {
                        name: "id",
                        optional: false,
                        structure: {
                          key: "Graph.apps.byId.arguments.id",
                          kind: "primitive",
                          type: "string",
                        },
                      },
                    ],
                    key: "Graph.apps.byId.arguments",
                    kind: "arguments",
                  },
                  key: "Graph.apps.byId",
                  kind: "function",
                  returns: {
                    key: "Graph.apps.byId.returns",
                    kind: "ref",
                    params: undefined,
                    ref: "App",
                  },
                },
              },
            ],
          },
        },
        {
          name: "flip",
          optional: false,
          structure: {
            key: "Graph.flip",
            kind: "union",
            types: [
              {
                key: "Graph.flip.union",
                kind: "ref",
                params: undefined,
                ref: "CoinHeads",
              },
              {
                key: "Graph.flip.union",
                kind: "ref",
                params: undefined,
                ref: "CoinTails",
              },
            ],
          },
        },
        {
          name: "config",
          optional: false,
          structure: {
            key: "Graph.config",
            kind: "ref",
            params: undefined,
            ref: "Config",
          },
        },
      ],
    },
    PaginationParams: {
      key: "PaginationParams",
      kind: "object",
      properties: [
        {
          name: "page",
          optional: true,
          structure: {
            key: "PaginationParams.page",
            kind: "primitive",
            type: "number",
          },
        },
        {
          name: "limit",
          optional: true,
          structure: {
            key: "PaginationParams.limit",
            kind: "primitive",
            type: "number",
          },
        },
      ],
    },
    Todo: {
      key: "Todo",
      kind: "object",
      properties: [
        {
          name: "todoName",
          optional: false,
          structure: {
            key: "Todo.todoName",
            kind: "primitive",
            type: "string",
          },
        },
        {
          name: "done",
          optional: false,
          structure: {
            key: "Todo.done",
            kind: "primitive",
            type: "boolean",
          },
        },
      ],
    },
    User: {
      key: "User",
      kind: "object",
      properties: [
        {
          name: "userName",
          optional: false,
          structure: {
            key: "User.userName",
            kind: "primitive",
            type: "string",
          },
        },
        {
          name: "app",
          optional: false,
          structure: {
            key: "User.app",
            kind: "ref",
            params: undefined,
            ref: "App",
          },
        },
      ],
    },
  },
}
`;

snapshot[`snapshot structure 1`] = `
{
  key: "",
  kind: "root",
  types: {
    Graph: {
      key: "Graph",
      kind: "object",
      properties: [
        {
          name: "group",
          optional: false,
          structure: {
            key: "Graph.group",
            kind: "ref",
            params: undefined,
            ref: "Group",
          },
        },
        {
          name: "user",
          optional: false,
          structure: {
            key: "Graph.user",
            kind: "ref",
            params: undefined,
            ref: "User",
          },
        },
      ],
    },
    Group: {
      key: "Group",
      kind: "object",
      properties: [
        {
          name: "name",
          optional: false,
          structure: {
            key: "Group.name",
            kind: "primitive",
            type: "string",
          },
        },
        {
          name: "users",
          optional: false,
          structure: {
            items: {
              key: "Group.users.items",
              kind: "ref",
              params: undefined,
              ref: "User",
            },
            key: "Group.users",
            kind: "array",
          },
        },
      ],
    },
    Role: {
      key: "Role",
      kind: "union",
      types: [
        {
          key: "Role.union",
          kind: "literal",
          type: "admin",
        },
        {
          key: "Role.union",
          kind: "literal",
          type: "user",
        },
      ],
    },
    User: {
      key: "User",
      kind: "object",
      properties: [
        {
          name: "name",
          optional: false,
          structure: {
            key: "User.name",
            kind: "primitive",
            type: "string",
          },
        },
        {
          name: "age",
          optional: false,
          structure: {
            key: "User.age",
            kind: "union",
            types: [
              {
                key: "User.age.union",
                kind: "primitive",
                type: "number",
              },
              {
                key: "User.age.union",
                kind: "literal",
                type: null,
              },
            ],
          },
        },
        {
          name: "group",
          optional: false,
          structure: {
            key: "User.group",
            kind: "ref",
            params: undefined,
            ref: "Group",
          },
        },
        {
          name: "maybeGroup",
          optional: false,
          structure: {
            key: "User.maybeGroup",
            kind: "union",
            types: [
              {
                key: "User.maybeGroup.union",
                kind: "ref",
                params: undefined,
                ref: "Group",
              },
              {
                key: "User.maybeGroup.union",
                kind: "literal",
                type: null,
              },
            ],
          },
        },
      ],
    },
  },
}
`;
